File: backend/app/api/v1/endpoints/leads.py
from typing import Any, List, Optional
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.orm import Session
from app import crud, models, schemas
from app.api import deps
from datetime import datetime
import uuid
from sqlalchemy.exc import IntegrityError
from pydantic import BaseModel
import logging
from sqlalchemy import func, and_, or_

logger = logging.getLogger(__name__)

router = APIRouter()

class LeadTagUpdate(BaseModel):
    tags: List[int]

class BulkLeadTagUpdate(BaseModel):
    leadIds: List[int]
    tags: List[int]

class LeadStatsResponse(BaseModel):
    total: int
    qualified: int
    hot_prospects: int
    qualification_rate: float

@router.get("/stats", response_model=LeadStatsResponse)
def get_lead_statistics(
    db: Session = Depends(deps.get_db),
    current_user: models.User = Depends(deps.get_current_user)
) -> Any:
    """
    Get lead statistics for the organization.
    """
    try:
        # Build organization filter
        filters = [
            models.Lead.is_deleted == False,
            models.Lead.visible == True
        ]
        
        # Add organization filter for non-admin users
        if not current_user.is_admin:
            filters.append(models.Lead.organization_id == current_user.organization_id)
        
        # Get total leads count
        total_leads = db.query(func.count(models.Lead.id)).filter(*filters).scalar() or 0
        
        # Get all leads for qualification analysis
        leads = db.query(models.Lead).filter(*filters).all()
        
        # Count qualified leads - more flexible criteria (at least email + (company or name))
        qualified_count = 0
        for lead in leads:
            has_email = bool(lead.email and lead.email.strip())
            has_company = bool(lead.company and lead.company.strip())
            has_name = bool((lead.first_name and lead.first_name.strip()) or (lead.last_name and lead.last_name.strip()))
            
            # Qualified if has email and either company or full name
            if has_email and (has_company or has_name):
                qualified_count += 1
        
        # Count hot prospects - more flexible criteria
        hot_sectors = ['technology', 'finance', 'healthcare', 'consulting', 'software', 'it', 'medical', 'tech', 'fintech']
        hot_sources = ['referral', 'linkedin', 'website', 'partner', 'recommendation']
        hot_prospects_count = 0
        
        for lead in leads:
            has_email = bool(lead.email and lead.email.strip())
            
            # Check if sector matches hot sectors
            is_hot_sector = False
            if lead.sector and lead.sector.strip():
                sector_lower = lead.sector.lower()
                is_hot_sector = any(hot_sector in sector_lower for hot_sector in hot_sectors)
            
            # Check if source is valuable
            is_hot_source = False
            if lead.source and lead.source.strip():
                source_lower = lead.source.lower()
                is_hot_source = any(hot_source in source_lower for hot_source in hot_sources)
            
            # Hot prospect if has email and (hot sector OR hot source OR has job title)
            has_job_title = bool(lead.job_title and lead.job_title.strip())
            
            if has_email and (is_hot_sector or is_hot_source or has_job_title):
                hot_prospects_count += 1
        
        # Calculate qualification rate
        qualification_rate = (qualified_count / total_leads * 100) if total_leads > 0 else 0.0
        
        logger.info(f"Lead Statistics for org {current_user.organization_id}: Total={total_leads}, Qualified={qualified_count}, Hot={hot_prospects_count}, Rate={qualification_rate}")
        
        return LeadStatsResponse(
            total=total_leads,
            qualified=qualified_count,
            hot_prospects=hot_prospects_count,
            qualification_rate=round(qualification_rate, 1)
        )
        
    except Exception as e:
        logger.error(f"Error retrieving lead statistics: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail=f"An error occurred while retrieving lead statistics: {str(e)}"
        )

@router.get("", response_model=List[schemas.Lead])
@router.get("/", response_model=List[schemas.Lead])
def read_leads(
    db: Session = Depends(deps.get_db),
    current_user: models.User = Depends(deps.get_current_user),
    skip: int = Query(0, ge=0),
    limit: int = Query(20, ge=1, le=100),
    search: Optional[str] = None,
    sort_by: Optional[str] = "created_at",
    sort_desc: bool = True,
    tag: Optional[str] = None
) -> Any:
    """
    Retrieve leads.
    Admin users can see all leads, normal users only see their organization's leads.
    """
    try:
        # Convert tag to tag_id if provided
        tag_id = None
        if tag and tag != 'all':
            try:
                tag_id = int(tag)
            except ValueError:
                raise HTTPException(
                    status_code=400,
                    detail="Invalid tag ID format"
                )

        # Get leads with organization filter
        leads = crud.lead.get_multi(
            db=db,
            skip=skip,
            limit=limit,
            organization_id=current_user.organization_id,
            search=search,
            sort_by=sort_by,
            sort_desc=sort_desc,
            tag_id=tag_id,
            is_admin=current_user.is_admin
        )
        
        # Artık psychometrics hybrid_property sayesinde otomatik olarak düzeltiliyor
        
        logger.info(f"Total leads before pagination: {len(leads)}")
        logger.info(f"Returning {limit} leads after pagination")
        
        return leads
        
    except Exception as e:
        logger.error(f"Error retrieving leads: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail=f"An error occurred while retrieving leads: {str(e)}"
        )

@router.post("", response_model=schemas.LeadResponse)
@router.post("/", response_model=schemas.LeadResponse)
def create_lead(
    *,
    db: Session = Depends(deps.get_db),
    current_user: models.User = Depends(deps.get_current_user),
    lead_in: schemas.LeadCreate,
) -> Any:
    """
    Create new lead.
    """
    try:
        # Log incoming data
        logger.info(f"Received lead creation request with data: {lead_in.model_dump()}")
        
        # Add organization_id, user_id, and created_by
        lead_data = lead_in.model_dump()
        lead_data["organization_id"] = current_user.organization_id
        lead_data["user_id"] = current_user.id
        lead_data["created_by"] = current_user.id
        lead_data["created_at"] = datetime.utcnow()
        lead_data["is_deleted"] = False
        lead_data["visible"] = True

        # Ensure psychometrics is a dict if provided
        if lead_data.get("psychometrics") == "":
            lead_data["psychometrics"] = None

        # Log enriched data
        logger.info(f"Enriched lead data: {lead_data}")

        # If stage_id is not provided, get or create default stage
        if not lead_data.get("stage_id"):
            logger.info("No stage_id provided, getting or creating default stage")
            default_stage = crud.lead_stage.get_by_order_index(
                db, 
                order_index=1, 
                organization_id=current_user.organization_id
            )
            if not default_stage:
                logger.info("Creating default stage")
                default_stage = crud.lead_stage.create(
                    db=db,
                    obj_in={
                        "name": "New",
                        "description": "Default stage for new leads",
                        "order_index": 1,
                        "organization_id": current_user.organization_id
                    }
                )
            lead_data["stage_id"] = default_stage.id
            logger.info(f"Set stage_id to {default_stage.id}")

        # Create lead
        try:
            lead = crud.lead.create(db=db, obj_in=lead_data)
            logger.info(f"Successfully created lead with ID: {lead.id}")
            
            # Handle tags if provided
            tags = []
            if lead_in.tags and len(lead_in.tags) > 0:
                logger.info(f"Adding tags to lead {lead.id}: {lead_in.tags}")
                
                # Add tags to the lead
                for tag_id in lead_in.tags:
                    tag = crud.tag.get(db=db, id=tag_id)
                    if tag and (current_user.is_admin or tag.organization_id == current_user.organization_id):
                        # Insert new association with organization_id
                        db.execute(
                            models.associations.lead_tags.insert().values(
                                lead_id=lead.id,
                                tag_id=tag_id,
                                organization_id=lead.organization_id,
                                created_at=datetime.utcnow()
                            )
                        )
                        tags.append({"id": tag.id, "name": tag.name})
                    else:
                        logger.warning(f"Tag {tag_id} not found or not accessible to user {current_user.id}")
                
                # Commit changes for tags
                db.commit()
                db.refresh(lead)
                logger.info(f"Successfully added tags to lead {lead.id}")
            
            # Convert lead object to dictionary
            lead_dict = {
                "id": lead.id,
                "first_name": lead.first_name,
                "last_name": lead.last_name,
                "company": lead.company,
                "job_title": lead.job_title,
                "email": lead.email,
                "telephone": lead.telephone,
                "mobile": lead.mobile,
                "location": lead.location,
                "linkedin": lead.linkedin,
                "country": lead.country,
                "website": lead.website,
                "sector": lead.sector,
                "organization_id": lead.organization_id,
                "user_id": lead.user_id,
                "stage_id": lead.stage_id,
                "created_by": lead.created_by,
                "created_at": lead.created_at,
                "is_deleted": lead.is_deleted,
                "visible": lead.visible,
                "psychometrics": lead.psychometrics or {},
                "notes": [],
                "tags": tags
            }
            
            return schemas.LeadResponse(
                success=True,
                message="Lead created successfully",
                data=lead_dict
            )
        except Exception as e:
            logger.error(f"Error creating lead in database: {str(e)}")
            raise

    except Exception as e:
        logger.error(f"Error in create_lead: {str(e)}")
        db.rollback()
        raise HTTPException(
            status_code=500,
            detail=f"Error creating lead: {str(e)}"
        )

@router.get("/{id}", response_model=schemas.LeadResponse)
@router.get("/{id}/", response_model=schemas.LeadResponse)
def read_lead(
    *,
    db: Session = Depends(deps.get_db),
    current_user: models.User = Depends(deps.get_current_user),
    id: int,
) -> Any:
    """
    Get lead by ID.
    """
    logger.info(f"Fetching lead with ID: {id} for user {current_user.id}")
    lead = crud.lead.get(db=db, id=id)
    if not lead:
        logger.warning(f"Lead with ID {id} not found")
        raise HTTPException(status_code=404, detail="Lead not found")
    
    # Admin users can access any lead
    if not current_user.is_admin:
        # Non-admin users can only access leads from their organization
        if lead.organization_id != current_user.organization_id:
            logger.warning(f"User {current_user.id} attempted to access lead {id} from organization {lead.organization_id}")
            raise HTTPException(status_code=403, detail="Not enough permissions")
    
    # Convert lead object to dictionary
    lead_dict = {
        "id": lead.id,
        "first_name": lead.first_name,
        "last_name": lead.last_name,
        "company": lead.company,
        "job_title": lead.job_title,
        "email": lead.email,
        "telephone": lead.telephone,
        "mobile": lead.mobile,
        "location": lead.location,
        "linkedin": lead.linkedin,
        "country": lead.country,
        "website": lead.website,
        "sector": lead.sector,
        "organization_id": lead.organization_id,
        "user_id": lead.user_id,
        "stage_id": lead.stage_id,
        "created_by": lead.created_by,
        "created_at": lead.created_at,
        "is_deleted": lead.is_deleted,
        "visible": lead.visible,
        "psychometrics": lead.psychometrics or {},
        "notes": [],
        "tags": [],
        "lab_comments": lead.lab_comments,
        "client_comments": lead.client_comments,
        "time_in_current_role": lead.time_in_current_role,
        "source": lead.source,
        "est_wealth_experience": lead.est_wealth_experience,
        "wpi": lead.wpi,
        "unique_lead_id": lead.unique_lead_id,
        "updated_at": lead.updated_at,
        "full_name": f"{lead.first_name} {lead.last_name}".strip()
    }

    # Add stage information if available
    if hasattr(lead, 'stage') and lead.stage:
        lead_dict["stage"] = {
            "id": lead.stage.id,
            "name": lead.stage.name
        }

    # Add user information if available
    if hasattr(lead, 'user') and lead.user:
        lead_dict["user"] = {
            "id": lead.user.id,
            "email": lead.user.email,
            "first_name": lead.user.first_name,
            "last_name": lead.user.last_name
        }

    # Add creator information if available
    if hasattr(lead, 'creator') and lead.creator:
        lead_dict["creator"] = {
            "id": lead.creator.id,
            "email": lead.creator.email,
            "first_name": lead.creator.first_name,
            "last_name": lead.creator.last_name
        }

    # Add tags if available
    if hasattr(lead, 'tags'):
        lead_dict["tags"] = [
            {
                "id": tag.id,
                "name": tag.name
            }
            for tag in lead.tags
        ]

    # Add notes if available
    if hasattr(lead, 'notes'):
        lead_dict["notes"] = [
            {
                "id": note.id,
                "content": note.content,
                "created_at": note.created_at,
                "created_by": note.created_by
            }
            for note in lead.notes
        ]
    
    return schemas.LeadResponse(
        success=True,
        message="Lead retrieved successfully",
        data=lead_dict
    )

@router.put("/{lead_id}", response_model=schemas.Lead)
def update_lead(
    *,
    db: Session = Depends(deps.get_db),
    lead_id: int,
    lead_in: schemas.LeadUpdate,
    current_user: models.User = Depends(deps.get_current_active_user),
) -> Any:
    """
    Update a lead.
    Admin users can update any lead, normal users can only update leads in their organization.
    """
    # Get the lead using the base class get method which expects 'id'
    lead = crud.lead.get(db=db, id=lead_id)
    if not lead:
        raise HTTPException(status_code=404, detail="Lead not found")
        
    # Check if user has permission to update this lead
    # Admin kullanıcılar herhangi bir lead'i güncelleyebilir
    # Normal kullanıcılar sadece kendi organizasyonlarındaki lead'leri güncelleyebilir
    if not current_user.is_admin and lead.organization_id != current_user.organization_id:
        logger.warning(f"Permission denied: User {current_user.id} attempted to update lead {lead_id} from organization {lead.organization_id}, but belongs to organization {current_user.organization_id}")
        raise HTTPException(status_code=403, detail="Not enough permissions")
        
    lead = crud.lead.update(db=db, db_obj=lead, obj_in=lead_in)
    logger.info(f"Lead {lead_id} successfully updated by user {current_user.id}")
    return lead

@router.post("/{lead_id}/tags", response_model=schemas.LeadResponse)
def update_lead_tags(
    lead_id: int,
    tag_update: LeadTagUpdate,
    db: Session = Depends(deps.get_db),
    current_user: models.User = Depends(deps.get_current_active_user)
) -> Any:
    """
    Update tags for a lead.
    Accepts a list of tag IDs and replaces all existing tags for the lead.
    """
    try:
        logger.info(f"Updating tags for lead {lead_id}: {tag_update.tags}")
        
        # Get the lead
        lead = crud.lead.get(db=db, id=lead_id)
        if not lead:
            raise HTTPException(status_code=404, detail="Lead not found")
        
        # Check permissions
        if not current_user.is_admin and lead.organization_id != current_user.organization_id:
            raise HTTPException(status_code=403, detail="Not enough permissions")
        
        # First, delete existing tag associations for this lead
        db.execute(
            models.associations.lead_tags.delete().where(
                models.associations.lead_tags.c.lead_id == lead_id
            )
        )
        
        # Add new tags
        for tag_id in tag_update.tags:
            tag = crud.tag.get(db=db, id=tag_id)
            if tag and (current_user.is_admin or tag.organization_id == current_user.organization_id):
                # Insert new association with organization_id
                db.execute(
                    models.associations.lead_tags.insert().values(
                        lead_id=lead_id,
                        tag_id=tag_id,
                        organization_id=lead.organization_id,
                        created_at=datetime.utcnow()
                    )
                )
            else:
                logger.warning(f"Tag {tag_id} not found or not accessible to user {current_user.id}")
        
        db.commit()
        db.refresh(lead)
        
        # Convert lead to dictionary for response
        lead_dict = {
            "id": lead.id,
            "first_name": lead.first_name,
            "last_name": lead.last_name,
            "tags": [{"id": tag.id, "name": tag.name} for tag in lead.tags]
        }
        
        logger.info(f"Successfully updated tags for lead {lead_id}")
        return schemas.LeadResponse(success=True, data=lead_dict)
    except Exception as e:
        logger.error(f"Error updating tags: {str(e)}")
        db.rollback()
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/bulk/tags", response_model=schemas.LeadResponse)
def update_leads_tags_bulk(
    tag_update: BulkLeadTagUpdate,
    db: Session = Depends(deps.get_db),
    current_user: models.User = Depends(deps.get_current_active_user)
) -> Any:
    """
    Update tags for multiple leads in bulk
    """
    try:
        if not tag_update.leadIds:
            raise HTTPException(
                status_code=400,
                detail="No leads provided for tagging"
            )

        if not tag_update.tags:
            raise HTTPException(
                status_code=400,
                detail="No tags provided for update"
            )

        # Verify all leads exist and user has permission
        leads = []
        for lead_id in tag_update.leadIds:
            lead = crud.lead.get(db=db, id=lead_id)
            if not lead:
                raise HTTPException(
                    status_code=404,
                    detail=f"Lead with id {lead_id} not found"
                )
            if lead.organization_id != current_user.organization_id:
                raise HTTPException(
                    status_code=403,
                    detail=f"Not enough permissions for lead {lead_id}"
                )
            leads.append(lead)

        # Verify all tags exist and belong to user's organization
        tags = []
        for tag_id in tag_update.tags:
            tag = crud.tag.get(db=db, id=tag_id)
            if not tag:
                raise HTTPException(
                    status_code=404,
                    detail=f"Tag with id {tag_id} not found"
                )
            if tag.organization_id != current_user.organization_id:
                raise HTTPException(
                    status_code=403,
                    detail=f"Not enough permissions for tag {tag_id}"
                )
            tags.append(tag)

        try:
            # First, delete existing tag associations for these leads
            for lead in leads:
                db.execute(
                    models.associations.lead_tags.delete().where(
                        models.associations.lead_tags.c.lead_id == lead.id
                    )
                )
            
            # Then create new associations
            for lead in leads:
                for tag in tags:
                    db.execute(
                        models.associations.lead_tags.insert().values(
                            lead_id=lead.id,
                            tag_id=tag.id,
                            organization_id=current_user.organization_id,
                            created_at=datetime.utcnow()
                        )
                    )
            
            # Commit all changes
            db.commit()
            
            logger.info(f"Successfully updated tags for {len(leads)} leads with tags: {[t.name for t in tags]}")
            
            return schemas.LeadResponse(
                success=True,
                data={
                    "updated_leads": len(leads),
                    "tags": [{"id": t.id, "name": t.name} for t in tags]
                }
            )
        except Exception as db_error:
            db.rollback()
            logger.error(f"Database error in bulk tag update: {str(db_error)}")
            raise HTTPException(
                status_code=500,
                detail="Failed to update tags due to database error"
            )
            
    except HTTPException as he:
        db.rollback()
        raise he
    except Exception as e:
        db.rollback()
        logger.error(f"Error in bulk tag update: {str(e)}")
        raise HTTPException(
            status_code=500,
            detail=f"Failed to update tags: {str(e)}"
        )

@router.delete("/{lead_id}", response_model=schemas.LeadResponse)
def delete_lead(
    *,
    db: Session = Depends(deps.get_db),
    current_user: models.User = Depends(deps.get_current_user),
    lead_id: int,
    confirm: bool = Query(False, description="Confirmation flag for lead deletion")
) -> Any:
    """
    Delete a lead.
    """
    lead = crud.lead.get(db=db, id=lead_id)
    if not lead:
        raise HTTPException(status_code=404, detail="Lead not found")
    if lead.organization_id != current_user.organization_id:
        raise HTTPException(status_code=403, detail="Not enough permissions")
    
    # If not confirmed, return a confirmation request
    if not confirm:
        return schemas.LeadResponse(
            success=False,
            message="Please confirm deletion",
            data={
                "id": lead_id,
                "first_name": lead.first_name,
                "last_name": lead.last_name,
                "company": lead.company,
                "requires_confirmation": True
            }
        )
    
    # If confirmed, proceed with deletion
    crud.lead.remove(db=db, id=lead_id)
    return schemas.LeadResponse(
        success=True,
        message="Lead deleted successfully",
        data={"id": lead_id}
    )

File: frontend/src/services/api/leads.ts
import api from '../axios';
// Lead duplicate tanımını ortadan kaldıralım
// import { Lead, LeadStage, LeadWithStats } from '@/types/leads';

export interface User {
  id: number;
  email: string;
  first_name: string;
  last_name: string;
}

export interface Stage {
  id: number;
  name: string;
}

export interface Tag {
  id: number;
  name: string;
  color: string;
}

export interface Note {
  id: number;
  content: string;
  created_at: string;
  updated_at: string;
  user: User;
}

export interface Lead {
  id: number;
  first_name: string;
  last_name: string;
  company: string | null;
  job_title: string | null;
  email: string | null;
  telephone: string | null;
  mobile: string | null;
  location: string | null;
  linkedin: string | null;
  country: string | null;
  website: string | null;
  sector: string | null;
  unique_lead_id: string | null;
  time_in_current_role: string | null;
  lab_comments: string | null;
  client_comments: string | null;
  psychometrics: Record<string, any> | null;
  wpi: string | null;
  source: string | null;
  est_wealth_experience: string | null;
  organization_id: number;
  user_id: number;
  stage_id: number;
  created_by: number;
  created_at: string;
  updated_at: string | null;
  is_deleted: boolean;
  visible: boolean;
  full_name: string;
  stage: Stage | null;
  user: User | null;
  creator: User | null;
  tags: Tag[];
  notes: Note[];
}

export interface LeadCreate {
  first_name: string;
  last_name: string;
  email?: string;
  company?: string;
  job_title?: string;
  location?: string;
  stage_id: number;
}

export interface LeadUpdate extends Partial<LeadCreate> {}

export interface LeadStage {
  id: number;
  name: string;
  description?: string;
  color: string;
  order_index: number;
  organization_id: number;
  is_active: boolean;
}

export interface LeadWithStats extends Lead {
  stats?: {
    total_deals?: number;
    total_value?: number;
    win_rate?: number;
    avg_deal_size?: number;
  };
}

export interface LeadStats {
  total: number;
  new: number;
  qualified: number;
  contacted: number;
  converted: number;
  hot_prospects?: number;
  qualification_rate?: number;
}

interface LeadsParams {
  search?: string;
  sort_by?: string;
  sort_desc?: boolean;
  stage_id?: number;
  assigned_to_id?: number;
  tag?: string;
  skip?: number;
  limit?: number;
}

interface LeadQueryParams {
  search?: string;
  skip?: number;
  limit?: number;
  sort_by?: string;
  sort_desc?: boolean;
}

interface LeadResponse {
  success: boolean;
  message: string;
  data: Lead | Lead[]; 
}

// Geliştirilmiş mock leads verisi
const mockLeads: Lead[] = Array.from({ length: 10 }, (_, i) => ({
  id: i + 1,
  first_name: `Lead ${i + 1}`,
  last_name: `Sample`,
  email: `lead${i + 1}@example.com`,
  company: `Company ${i + 1}`,
  job_title: `Position ${i + 1}`,
  telephone: null,
  mobile: null,
  location: null,
  linkedin: null, 
  country: null,
  website: null,
  sector: null,
  unique_lead_id: `LEAD-${i + 1000}`,
  time_in_current_role: null,
  lab_comments: null,
  client_comments: null,
  psychometrics: null,
  wpi: null,
  source: null,
  est_wealth_experience: null,
  organization_id: 1,
  user_id: 1,
  stage_id: 1,
  created_by: 1,
  created_at: new Date().toISOString(),
  updated_at: new Date().toISOString(),
  is_deleted: false,
  visible: true,
  full_name: `Lead ${i + 1} Sample`,
  stage: { id: 1, name: 'New' },
  user: { id: 1, email: 'admin@example.com', first_name: 'Admin', last_name: 'User' },
  creator: { id: 1, email: 'admin@example.com', first_name: 'Admin', last_name: 'User' },
  tags: [],
  notes: []
}));

// Her zaman response verecek şekilde API çağrıları
export const getLeads = async (params: LeadQueryParams = {}): Promise<any> => {
  console.log('Getting leads with params:', params);
  try {
    // URL sonuna / ekleyerek yönlendirmeyi önleyelim
    const response = await api.get('/leads/', { params });
    console.log('Leads API response:', response.data);
    return response.data;
  } catch (error) {
    console.log('Error in getLeads:', error);
    throw error;
  }
};

export const getById = async (id: number): Promise<LeadResponse> => {
  try {
    const response = await api.get(`/leads/${id}/`);
    return response.data;
  } catch (error) {
    console.log('Error in getById:', error);
    throw error;
  }
};

export const create = async (data: Partial<Lead>): Promise<LeadResponse> => {
  try {
    const response = await api.post(`/leads/`, data);
    return response.data;
  } catch (error) {
    console.error('Error creating lead:', error);
    throw error;
  }
};

export const update = async (id: number, data: Partial<Lead>): Promise<LeadResponse> => {
  try {
    const response = await api.put(`/leads/${id}/`, data);
    return response.data;
  } catch (error) {
    console.error('Error updating lead:', error);
    throw error;
  }
};

export const remove = async (id: number): Promise<LeadResponse> => {
  try {
    const response = await api.delete(`/leads/${id}/`);
    return response.data;
  } catch (error) {
    console.error('Error deleting lead:', error);
    throw error;
  }
};

const updateTags = async (id: number, data: { tags: number[] }) => {
  try {
    const response = await api.patch<Lead>(`/leads/${id}/tags/`, data);
    return response.data;
  } catch (error) {
    console.error('Error updating lead tags:', error);
    throw error;
  }
};

// leadsAPI export'u ekleniyor
export const leadsAPI = {
  getAll: async (params?: { 
    search?: string;
    sort_by?: string;
    sort_desc?: boolean;
    stage_id?: number;
    assigned_to_id?: number;
  }): Promise<any> => {
    return api.get('/leads/', { params });
  },

  getLeads: async (params?: { 
    search?: string;
    sort_by?: string;
    sort_desc?: boolean;
    stage_id?: number;
    assigned_to_id?: number;
    skip?: number;
    limit?: number;
  }): Promise<any> => {
    const response = await api.get('/leads/', { params });
    const data = response.data;
    // Backend returns array, wrap it in expected format
    return {
      results: Array.isArray(data) ? data : data.results || [],
      total: Array.isArray(data) ? data.length : data.total || 0,
      page: params?.skip ? Math.floor(params.skip / (params.limit || 20)) : 0,
      limit: params?.limit || 20
    };
  },

  getById: async (id: number): Promise<any> => {
    return api.get(`/leads/${id}`);
  },

  create: async (data: Partial<Lead>): Promise<any> => {
    return api.post('/leads/', data);
  },

  update: async (id: number, data: Partial<Lead>): Promise<any> => {
    return api.put(`/leads/${id}`, data);
  },

  delete: async (id: number): Promise<any> => {
    return api.delete(`/leads/${id}`);
  },

  uploadCSV: async (formData: FormData): Promise<any> => {
    return api.post('/leads/import/csv', formData, {
      headers: {
        'Content-Type': 'multipart/form-data',
      },
    });
  },

  exportCSV: async (params?: {
    search?: string;
    tag?: string;
    filters?: any;
  }): Promise<Blob> => {
    const response = await api.get('/leads/export/csv', {
      params,
      responseType: 'blob',
    });
    return response.data;
  },

  downloadTemplate: async (): Promise<Blob> => {
    const response = await api.get('/leads/template', {
      responseType: 'blob',
    });
    return response.data;
  },

  requestInfo: async (data: { lead_id: number; field_name: string; notes?: string }): Promise<any> => {
    return await api.post('/requests/info-request', data);
  },

  getLeadStats: async (): Promise<LeadStats> => {
    const response = await api.get('/leads/stats');
    return response.data;
  }
};

export default leadsAPI;
